File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.java
import java.io.*;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class Main {

    static int N;
    static ArrayList<Integer>[] graph;
    static boolean[] visited;
    static int[] answer;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(br.readLine()); //노드 수

        answer = new int[N + 1];
        visited = new boolean[N + 1];
        graph = new ArrayList[N + 1];
        for (int j = 0; j <= N; j++) {
            graph[j] = new ArrayList<>();
        }


        for (int i = 0; i < N - 1; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int x = Integer.parseInt(st.nextToken());
            int y = Integer.parseInt(st.nextToken());

            graph[x].add(y);
            graph[y].add(x);
        }

        dfs(1, 0);

        for (int i = 2; i <= N; i++) {
            System.out.println(answer[i]);
        }
        br.close();
    }

    private static void dfs(int idx, int before) {
        visited[idx] = true;
        answer[idx] = before;

        for (int i = 0; i < graph[idx].size(); i++) {
            if (visited[graph[idx].get(i)] == false) {
                dfs(graph[idx].get(i), idx);
            }
        }
    }

}

File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.py
def dfs(v):
    # dfs 할때마다 result에 각 노드를 추가한다.
    visited[v] = True

    for next in graph[v]:
        if not visited[next]:
            dfs(next)
            answer[next] = v

import sys
sys.setrecursionlimit(10 ** 6)
N = int(input())

# 인접 리스트를 만든다.
graph = [[] for i in range(N + 1)]
answer = [0] * (N + 1)
visited =[False] * (N + 1)

# 인접 리스트를 입력받는다.
for i in range(N-1):
    start, end = map(int, sys.stdin.readline().split())
    graph[start].append(end)
    graph[end].append(start)


# 루트 노드가 1이므로, 1부터 시작해서 dfs를 시작한다.
dfs(1)

for i in range(2, N+1):
    print(answer[i])